explorer/interpreter/type_checker.cpp

```diff
@@ -649,6 +780,22 @@ auto TypeChecker::ArgumentDeduction(
                                           << "actual: " << *arg;
     }
     const Value* subst_param_type = Substitute(deduced, param);
+    if (param->kind() == Value::Kind::VariableType) {
```

Comment:
I think this logic would fit better in the `case VariableType` code below rather than in `handle_non_deduced_type`, because I think we want to use it for all `VariableType` cases regardless of whether the parameter type is deduced. Perhaps we could do this check at the start of that `case`, on line 808?